### PR TITLE
 drivers/lis2dh12: clean up API

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Previously the lis2dh12 driver was mostly just exposing raw registers to the user and push interpreting the register data to the application.
This makes for convoluted and error prone application code, so instead handle all the register writing inside the driver and expose a more high-level interface to the user.

This also means we no longer expose the raw interrupt callback to the user. This only leads to the user trying to access the I2C/SPI bus inside interrupt context (as done by the test application) to read the interrupt source.
This is against the API contract and can lead to lock-ups if the user process is accessing the bus at the same time.

Instead this introduces a blocking `lis2dh12_wait_event()` function that returns a bit map of the interrupt source once it occurs. 

### Testing procedure

For easier use the single `lis` command in the test application was split into multiple separate shell commands.

to configure a click event (1000 mg for less than 100ms)

    click 1000 100000

for a threshold event (2000 mg)

    treshold 1 2000

Tap the device

```
2021-04-13 17:59:34,790 #  event 1
2021-04-13 17:59:34,791 # event 1
2021-04-13 17:59:34,794 # event 1
2021-04-13 17:59:34,795 # event 1
2021-04-13 17:59:34,797 # event 1
2021-04-13 17:59:35,789 # event 1
2021-04-13 17:59:35,791 # event 1
2021-04-13 17:59:35,792 # event 1
2021-04-13 17:59:35,794 # event 1
2021-04-13 17:59:35,796 # event 1
2021-04-13 17:59:36,102 # event 1
2021-04-13 17:59:36,104 # event 1
2021-04-13 17:59:36,106 # event 1
2021-04-13 17:59:36,108 # event 1
2021-04-13 17:59:36,109 # event 1
2021-04-13 17:59:36,592 # event 1
2021-04-13 17:59:36,594 # event 1
2021-04-13 17:59:36,596 # event 1
2021-04-13 17:59:36,598 # event 1
2021-04-13 17:59:36,601 # event 1
2021-04-13 17:59:37,288 # event 1
2021-04-13 17:59:37,291 # event 1
2021-04-13 17:59:37,292 # event 1
2021-04-13 17:59:37,296 # event 1
2021-04-13 17:59:37,300 # event 1
2021-04-13 17:59:37,720 # click event
2021-04-13 17:59:37,722 # click event
2021-04-13 17:59:37,724 # click event
```

### Issues/PRs references

CC @jan-mo 

follow-up on #13472 & #15871

### CODE REVIEW

**Comment by fjmolinas** [comment]
Init fails on `dwm1001`, but it's also the case in master...

File: drivers/include/lis2dh12.h [hunk 1]
```diff
@@ -239,33 +244,71 @@ typedef struct {
...
+ * @param[in] mg        acceleration in mg
+ * @param[in] us_limit  upper limit for click duration in µs
+ * @param[in] us_latency  dead time after click event in µs
+ * @param[in] us_window time after @p us_latency in which the second click event
+ *                      must occur to register as double click
+ * @param[in] click     bit map of click axis / types
+ * @param[in] pin       Interrupt pin to use (LIS2DH12_INT1/LIS2DH12_INT2)
  */
-int lis2dh12_set_int(const lis2dh12_t *dev, const lis2dh12_int_params_t *params, uint8_t int_line);
+void lis2dh12_cfg_click_event(const lis2dh12_t *dev, uint32_t mg,
```
Comment (fjmolinas) [inline drivers/include/lis2dh12.h:286]: Is this type of event common application for this device? It seems somewhat specific.

Comment (benpicco) [reply drivers/include/lis2dh12.h:286]: It's a hardware feature

image

**Comment by benpicco** [comment]
> Init fails on `dwm1001`, but it's also the case in master...

try with `DRIVER=lis2dh12_i2c`

**Comment by fjmolinas** [comment]
> > Init fails on `dwm1001`, but it's also the case in master...
> 
> try with `DRIVER=lis2dh12_i2c`

Thanks now it works, some things when testing:

- multiple click events for every tap (about 5)
- temperature seems a bit off

**Comment by fjmolinas** [comment]
> This also means we no longer expose the raw interrupt callback to the user. This only leads to the user trying to access the I2C/SPI bus inside interrupt context (as done by the test application) to read the interrupt source.
> This is against the API contract and can lead to lock-ups if the user process is accessing the bus at the same time.

I have just took a quick look through the datasheet so please correct me if I'm wrong, but doesn't the API change limit the possible interrupt souces the user can set with `lis2dh12_set_int` gone?

**Comment by benpicco** [comment]
There are only 3 possible interrupt sources

 - INT1 event
 - INT2 event
 - click event

Those can be configured with `lis2dh12_cfg_threshold_event()` for INT1 & INT2 and `lis2dh12_cfg_click_event()` for the click event.

`lis2dh12_wait_event()` is then used to wait for any of these events.

**Comment by fjmolinas** [comment]
> There are only 3 possible interrupt sources
> 
>     * INT1 event
> 
>     * INT2 event
> 
>     * click event
> 
> 
> Those can be configured with `lis2dh12_cfg_threshold_event()` for INT1 & INT2 and `lis2dh12_cfg_click_event()` for the click event.
> 
> `lis2dh12_wait_event()` is then used to wait for any of these events.

Ok, and what about the click events, do you also get multiple event for every tap?

**Comment by benpicco** [comment]
> Ok, and what about the click events, do you also get multiple event for every tap?

depends on how much your device will bounce and what thresholds you have set, typically it should be only one event though.

image

with double click you only get an event every second tap

image

**Comment by fjmolinas** [comment]
> depends on how much your device will bounce and what thresholds you have set, typically it should be only one event though.

I used the values described in the test procedure.

**Comment by fjmolinas** [comment]
> > depends on how much your device will bounce and what thresholds you have set, typically it should be only one event though.
> 
> I used the values described in the test procedure.

From the test procedure, I see you get the same results, every event is received at least 3 times which seems wrong.

```
2021-04-13 17:59:37,720 # click event
2021-04-13 17:59:37,722 # click event
2021-04-13 17:59:37,724 # click event
```

And I have the same experience with the `threshold event` it's also generated ~4 times for every event, and I've tried playing around with the values with no success:

```
2021-05-21 12:00:37,500 # event 1
2021-05-21 12:00:37,502 # event 1
2021-05-21 12:00:37,504 # event 1
2021-05-21 12:00:38,605 # event 1
2021-05-21 12:00:38,606 # event 1
2021-05-21 12:00:38,608 # event 1
2021-05-21 12:00:38,610 # event 1
2021-05-21 12:00:38,613 # event 1
```

**Comment by benpicco** [comment]
Ok, this is due to the stale event reporting - the CPU is faster than the sensor at updating the register.
I've added an option to ignore the stale events.

The stale events are needed if the interrupt is used to wake the CPU from Deep Sleep - here we *want* to read the old event that caused the wake-up.

File: drivers/include/lis2dh12.h [hunk 2]
```diff
@@ -239,33 +244,74 @@ typedef struct {
...
+
+/**
+ * @brief   Wait for an interrupt event
+ *          This function will block until an interrupt is received
  *
  * @param[in] dev      device descriptor
- * @param[out] data    device interrupt data
- * @param[in] int_line number of interrupt line (LIS2DH12_INT1 or LIS2DH12_INT2)
+ * @param[in] pin      Interrupt pin to monitor (LIS2DH12_INT1 or LIS2DH12_INT2)
+ * @param[in] state_events  If true, this also reports events that were generated
```
Comment (fjmolinas) [inline drivers/include/lis2dh12.h:307]:
```suggestion
 * @param[in] stale_events  If true, this also reports events that were generated
```

**Comment by fjmolinas** [comment]
@benpicco please squash after fixing the static-check issue.

**Review by fjmolinas — APPROVED** [review]
ACK, events seem precise enough:

```
2021-06-03 13:20:54,580 # click 1000 50000
2021-06-03 13:20:54,585 # error: 0
> 2021-06-03 13:20:55,498 # click event
2021-06-03 13:20:56,250 # click event
2021-06-03 13:20:57,503 # click event
2021-06-03 13:20:58,649 # click event
threshold 1 2000
```

```
2021-06-03 13:21:14,505 # main(): This is RIOT! (Version: 2021.04-devel-1341-ga338b-pr-16328)
2021-06-03 13:21:14,510 # using I2C mode, for SPI mode select the lis2dh12_spi module
2021-06-03 13:21:14,517 # lis2dh12 [Initialized]
threshold 1 2000
2021-06-03 13:21:16,747 # threshold 1 2000
> 2021-06-03 13:21:19,760 # error: 0
2021-06-03 13:21:21,746 # event 1
2021-06-03 13:21:21,784 # event 1
2021-06-03 13:21:21,813 # event 1
2021-06-03 13:21:22,662 # event 1
2021-06-03 13:21:22,699 # event 1
2021-06-03 13:21:22,728 # event 1
2021-06-03 13:21:23,528 # event 1
2021-06-03 13:21:23,556 # event 1
2021-06-03 13:21:24,203 # event 1
2021-06-03 13:21:24,751 # event 1
2021-06-03 13:21:25,320 # event 1
2021-06-03 13:21:25,754 # event 1
2021-06-03 13:21:26,217 # event 1
2021-06-03 13:21:26,669 # event 1
2021-06-03 13:21:27,420 # event 1
2021-06-03 13:21:27,941 # event 1
2021-06-03 13:21:27,989 # event 1
2021-06-03 13:21:30,137 # event 1
2021-06-03 13:21:30,166 # event 1
2021-06-03 13:21:30,947 # event 1
2021-06-03 13:21:30,994 # event 1
2021-06-03 13:21:31,023 # event 1
2021-06-03 13:21:31,997 # event 1
2021-06-03 13:21:32,767 # event 1
2021-06-03 13:21:33,528 # event 1
2021-06-03 13:21:34,270 # event 1
2021-06-03 13:21:35,581 # event 1
2021-06-03 13:21:36,612 # event 1
2021-06-03 13:21:36,640 # event 1
2021-06-03 13:21:36,678 # event 1
2021-06-03 13:21:36,707 # event 1
```

**Pull request merged**